UT/Examples for resharding checkpoint save/loads for distributed tensors with uneven shards.

pradeepfn opened the pull request:
1\  DTensor abstraction on its own, does not support arbitrary length shards in its distributed tensors representation. It supports a single uneven shard, bit it has to be the last shard in the sharding dimension.

2\ However, DCP supports an API called checkpointable. This API allows you to define your custom shardable tensor structure. I have given a UT example ( look for CheckpointableDistTensor). Therefore, one option is to use CheckpointableDistTensor to save/load uneven shards.

3\ While exploring this path, I also noticed that torch.rec module also encountered a similar problem while working with DTensor. They workaround it by implementing Checkpointable API in DTensor and introducing an auxillary structure called LocalShardsWrapper. This is the second option we can use to unblock data loader resharding work.

In summary;
Use LocalShardWrapper + DTensor as the first option to unblock.
Second preference is to use new implementation of Checkpointable API. ( similar to CheckpointbaleDistTensor I have introduced in this example).

Differential Revision: D80182564




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

XilunWu commented:
Quick question: any reason that we do not reuse [`LocalShardsWrapper`](https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_shards_wrapper.py#L27C7-L27C25)? I feel this is a good representation for arbitrary length shards. Besides, DTensor already implements `_Checkpointable` protocol. See: https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/planner_helpers.py#L324-L326

XilunWu commented:
1D device mesh -> [ 1, 2, 3, 4]
Distributed tensor (1d tensor with the values) I want to checkpoint -> [ 1, 2, 3, 4 ,5, 6]
Shard placement and I want among the devices ->
dev 1 -> [ 1,2]
dev 2 -> [ 3 ]
dev 3 -> [ 4, 5]
dev 4 -> [ 6 ]

A DTensor + LocalShardsWrapper representation is:
dev 1 -> local_tensor=LocalShardsWrapper(local_shards=[1, 2], local_offsets=[0])
dev 2 -> local_tensor=LocalShardsWrapper(local_shards=[3], local_offsets=[2])
dev 3 -> local_tensor=LocalShardsWrapper(local_shards=[4, 5], local_offsets=[3])
dev 4 -> local_tensor=LocalShardsWrapper(local_shards=[6], local_offsets=[5])

Then wrap it inside DTensor on every device:
distributed_checkpointable_dtensor = DTensor.from_local(
    local_tensor=local_tensor,
    device_mesh=...,
    placements=[Shard(0)],  # 0-dim sharding
    shape=torch.Size((6,)),  # the overall_size you want to set
    stride=(1,),  # stride
)

Back to your representation:
CheckpointableDistTensor._local_tensor == local_tensor
CheckpointableDistTensor._shard_offsets == local_tensor.local_offsets  # if you need all shards' offsets, need all-gather them
CheckpointableDistTensor._overall_size == distributed_checkpointable_dtensor.shape

pradeepfn commented:
> 1D device mesh -> [ 1, 2, 3, 4] Distributed tensor (1d tensor with the values) I want to checkpoint -> [ 1, 2, 3, 4 ,5, 6] Shard placement and I want among the devices -> dev 1 -> [ 1,2] dev 2 -> [ 3 ] dev 3 -> [ 4, 5] dev 4 -> [ 6 ]
> 
> A DTensor + LocalShardsWrapper representation is: dev 1 -> local_tensor=LocalShardsWrapper(local_shards=[1, 2], local_offsets=[0]) dev 2 -> local_tensor=LocalShardsWrapper(local_shards=[3], local_offsets=[2]) dev 3 -> local_tensor=LocalShardsWrapper(local_shards=[4, 5], local_offsets=[3]) dev 4 -> local_tensor=LocalShardsWrapper(local_shards=[6], local_offsets=[5])
> 
> Then wrap it inside DTensor on every device: distributed_checkpointable_dtensor = DTensor.from_local( local_tensor=local_tensor, device_mesh=..., placements=[Shard(0)], # 0-dim sharding shape=torch.Size((6,)), # the overall_size you want to set stride=(1,), # stride )
> 
> Back to your representation: CheckpointableDistTensor._local_tensor == local_tensor CheckpointableDistTensor._shard_offsets == local_tensor.local_offsets # if you need all shards' offsets, need all-gather them CheckpointableDistTensor._overall_size == distributed_checkpointable_dtensor.shape

Cool. I this fits the bill. Will add UT and verify.

pradeepfn committed:
[RFC] Supporting resharding checkpoint save/loads for distributed tensors with uneven shards. (#160533)

Summary:

1\  DTensor abstraction on its own, does not support arbitrary length shards in its distributed tensors representation. It supports a single uneven shard, bit it has to be the last shard in the sharding dimension.

2\ However, DCP supports an API called checkpointable. This API allows you to define your custom shardable tensor structure. I have given a UT example ( look for CheckpointableDistTensor). Therefore, one option is to use CheckpointableDistTensor to save/load uneven shards.

3\ While exploring this path, I also noticed that torch.rec module also encountered a similar problem while working with DTensor. They workaround it by implementing Checkpointable API in DTensor and introducing an auxillary structure called LocalShardsWrapper. This is the second option we can use to unblock data loader resharding work.



In summary;

1. Use LocalShardWrapper + DTensor as the first option to unblock.
2. Second preference is to use new implementation of Checkpointable API. ( similar to CheckpointbaleDistTensor I have introduced in this example).

Test Plan:
buck test  'fbcode//mode/dev-nosan' fbcode//caffe2/test/distributed/checkpoint:test_dtensor_resharding -- --exact 'caffe2/test/distributed/checkpoint:test_dtensor_resharding - test_dtensor_resharding.TestCheckpointableReshard: test_uneven_reshard_with_dtensor_shards_wrapper_api'








[pradeepfdo@devvm14571.frc0 ~/fbsource/fbcode (ccaa6cb220)]$ buck test  'fbcode//mode/dev-nosan' fbcode//caffe2/test/distributed/checkpoint:test_dtensor_resharding -- --exact 'caffe2/test/distributed/checkpoint:test_dtensor_resharding - test_dtensor_resharding.TestCheckpointableReshard: test_uneven_reshard_with_dtensor_shards_wrapper_api'
Buck UI: https://www.internalfb.com/buck2/fd32ce3a-de9d-486a-962b-9f9acb6b4667
Test UI: https://www.internalfb.com/intern/testinfra/testrun/6473924764240103
Network: Up: 36KiB  Down: 139KiB  (reSessionID-493a2cab-5791-4c34-93e5-e1dec39edff9)
Command: test.
Time elapsed: 1:09.3s
Tests finished: Pass 2. Fail 0. Fatal 0. Skip 0. Build failure 0
[pradeepfdo@devvm14571.frc0 ~/fbsource/fbcode (ccaa6cb220)]$

Differential Revision: D80182564

pradeepfn commented:
@pytorchbot merge
